xde/src/xde.rs

```diff
@@ -2071,26 +2067,59 @@ unsafe extern "C" fn xde_rx(
             mp_chain as uintptr_t,
             c"rx'd packet chain was null",
         );
-        return;
+        return ptr::null_mut();
     };
 
+    let mut out_chain = MsgBlkChain::empty();
+    let mut count = 0;
+    let mut len = 0;
+
     // TODO: In future we may want to batch packets for further tx
     // by the mch they're being targeted to. E.g., either build a list
     // of chains (port0, port1, ...), or hold tx until another
     // packet breaks the run targeting the same dest.
     while let Some(pkt) = chain.pop_front() {
         unsafe {
-            xde_rx_one(&stream, mrh, pkt);
+            if let Some(pkt) = xde_rx_one(&stream, ptr::null_mut(), pkt) {
+                count += 1;
+                len += pkt.byte_len();
+                out_chain.append(pkt);
+            }
         }
     }
+
+    let (head, tail) = out_chain
+        .unwrap_mblk_and_tail()
+        .map(|v| (v.0.as_ptr(), v.1.as_ptr()))
+        .unwrap_or((ptr::null_mut(), ptr::null_mut()));
+
+    if let Some(ptr) = NonNull::new(out_len) {
+        unsafe {
+            ptr.write(len);
+        }
+    }
+
+    if let Some(ptr) = NonNull::new(out_count) {
+        unsafe {
+            ptr.write(count);
+        }
+    }
+
+    if let Some(ptr) = NonNull::new(out_mp_tail) {
+        unsafe {
+            ptr.write(tail);
+        }
+    }
+
+    head
 }
 
 #[inline]
 unsafe fn xde_rx_one(
     stream: &DlsStream,
     mrh: *mut mac::mac_resource_handle,
     mut pkt: MsgBlk,
-) {
+) -> Option<MsgBlk> {
```

Comment:
Please document the expectations of the return value ("non-None when packet was not of interest to xde" or whatever)